lib/src/solver/failure.dart

```diff
@@ -19,6 +23,294 @@ class SolveFailure implements ApplicationException {
     assert(incompatibility.terms.single.package.isRoot);
   }
 
-  // TODO(nweiz): Produce a useful error message.
-  String toString() => "Tough luck, Chuck!";
+  /// Describes how [incompatibility] was derived, and thus why version solving
+  /// failed.
+  ///
+  /// See https://github.com/dart-lang/pub/tree/master/doc/solver.md#error-reporting
+  /// for details on how this algorithm works.
+  String toString() {
+    var derivations = _countDerivations();
+
+    // The lines in the proof. Each line is a message/number pair. The message
+    // describes a single incompatibility, and why its terms are incompatible.
+    // The number is optional and indicates the explicit number that should be
+    // associated with the line so it can be referred to later on.
+    var lines = <Pair<String, int>>[];
+
+    // A map from incompatibilities to the line numbers that were written for
+    // those incompatibilities.
+    var lineNumbers = <Incompatibility, int>{};
+
+    // Writes [message] to [lines].
+    //
+    // The [message] should describe [incompatibility] and how it was derived
+    // (if applicable). If [numbered] is true, this will associate a line number
+    // with [incompatibility] and [message] so that the message can be easily
+    // referred to later.
+    void write(Incompatibility incompatibility, String message,
+        {bool numbered: false}) {
+      if (numbered) {
+        var number = lineNumbers.length + 1;
+        lineNumbers[incompatibility] = number;
+        lines.add(new Pair(message, number));
+      } else {
+        lines.add(new Pair(message, null));
+      }
+    }
+
+    /// Returns whether we can collapse the derivation of [incompatibility].
+    ///
+    /// If [incompatibility] is only used to derive one other incompatibility,
+    /// it may make sense to skip that derivation and just derive the second
+    /// incompatibility directly from three causes. This is usually clear enough
+    /// to the user, and makes the proof much terser.
```

Comment:
Can you give an example? (If that's not easy, feel free to ignore this.)

Reply:
Done.